fem/bilininteg_ext.cpp

```diff
@@ -0,0 +1,1581 @@
+               const double gY = grad[qz][qy][qx][1];
+               const double gZ = grad[qz][qy][qx][2];
+               for (int dx = 0; dx < D1D; ++dx)
+               {
+                  const double wx  = Bt(dx,qx);
+                  const double wDx = Gt(dx,qx);
+                  gradX[dx][0] += gX * wDx;
+                  gradX[dx][1] += gY * wx;
+                  gradX[dx][2] += gZ * wx;
+               }
+            }
+            for (int dy = 0; dy < D1D; ++dy)
+            {
+               const double wy  = Bt(dy,qy);
+               const double wDy = Gt(dy,qy);
+               for (int dx = 0; dx < D1D; ++dx)
+               {
+                  gradXY[dy][dx][0] += gradX[dx][0] * wy;
+                  gradXY[dy][dx][1] += gradX[dx][1] * wDy;
+                  gradXY[dy][dx][2] += gradX[dx][2] * wy;
+               }
+            }
+         }
+         for (int dz = 0; dz < D1D; ++dz)
+         {
+            const double wz  = Bt(dz,qz);
+            const double wDz = Gt(dz,qz);
+            for (int dy = 0; dy < D1D; ++dy)
+            {
+               for (int dx = 0; dx < D1D; ++dx)
+               {
+                  y(dx,dy,dz,e) +=
+                     ((gradXY[dy][dx][0] * wz) +
+                      (gradXY[dy][dx][1] * wz) +
+                      (gradXY[dy][dx][2] * wDz));
+               }
+            }
+         }
+      }
+   });
+}
+
+typedef void (*fDiffusionMultAdd)(const int NE,
+                                  const double* B,
+                                  const double* G,
+                                  const double* Bt,
+                                  const double* Gt,
+                                  const double* oper,
+                                  const double* solIn,
+                                  double* solOut);
+
+static void PADiffusionMultAssembled(const int dim,
+                                     const int D1D,
+                                     const int Q1D,
+                                     const int NE,
+                                     const double* B,
+                                     const double* G,
+                                     const double* Bt,
+                                     const double* Gt,
+                                     const double* op,
+                                     const double* x,
+                                     double* y)
+{
+#ifdef MFEM_USE_OCCA
+   if (Device::UsingOcca())
+   {
+      MFEM_ASSERT(dim == 2, "");
+      OccaPADiffusionMultAdd2D(D1D, Q1D, NE, B, G, Bt, Gt, op, x, y);
+      return;
+   }
+#endif // MFEM_USE_OCCA
+   const int id = (dim<<8)|(D1D<<4)|(Q1D);
+   static std::unordered_map<int, fDiffusionMultAdd> call =
+   {
+      // 2D
+      {0x222,&PADiffusionMultAssembled2D<2,2>},
+      {0x233,&PADiffusionMultAssembled2D<3,3>},
+      {0x244,&PADiffusionMultAssembled2D<4,4>},
+      // 3D
+      {0x323,&PADiffusionMultAssembled3D<2,3>},
```

Comment:
It seems we can't run out of the box with order > 1 in 3D. What's the plan for higher orders? Would it be possible to have several optimized instantiations with known sizes (for common orders like 1, 2, 3), and a general, run-time size fallback? BTW, how much slower is a kernel where the sizes are not known at compile time?

Reply:
You're right, I haven't enabled most of the 3D due to compilation time, but we should re-enable them for the common orders. The fall-back should be available with the generic kernels (#758).

Reply:
Maybe we should make a list of all the orders we want by default, and add an error message that explain the user how to modify the code to add the order of interest. Unfortunately this procedure can be a bit obscure since the number of quadrature points used is difficult to track, so giving directly to the user the line that he should add we be the best I think, _i.e._ `PADiffusionMultAssembled kernel not instanciated, Add the kernel "0xDimNDofNQuad,&PADiffusionMultAssembledDimD<NDof,NQuad>" ` with `Dim`,`NDof` and `NQuad` having the desired value to make adding a kernel as effortless as possible.